src/web/services/search.service.ts

```diff
@@ -0,0 +1,127 @@
+import { ResourceEndpoints } from '../types/api-endpoints';
+import { InstructorPrivilege, Student, Students } from '../types/api-output';
+import { HttpRequestService } from './http-request.service';
+
+/**
+ * Handles the logic for search.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class SearchService {
+  constructor(private httpRequestService: HttpRequestService) {}
+
+  searchInstructor(searchKey: string): Observable<InstructorSearchResult> {
+    return this.getStudents(searchKey).pipe(
+      map((studentsRes: Students) => this.getCoursesWithSections(studentsRes)),
+      mergeMap((coursesWithSections: SearchStudentsTable[]) =>
+        forkJoin([
+          of(coursesWithSections),
+          this.getPrivileges(coursesWithSections),
+        ]),
+      ),
+      map((res: [SearchStudentsTable[], InstructorPrivilege[]]) => this.combinePrivileges(res)),
+    );
+  }
+
+  getStudents(searchKey: string): Observable<Students> {
+    const paramMap: { [key: string]: string } = {
+      searchkey: searchKey,
+    };
+    return this.httpRequestService.get('/search/students', paramMap);
+  }
+
+  getCoursesWithSections(studentsRes: Students): SearchStudentsTable[] {
+    const { students }: { students: Student[] } = studentsRes;
+
+    const distinctCourses: string[] = Array.from(
+      new Set(students.map((s: Student) => s.courseId)),
+    );
+    const coursesWithSections: SearchStudentsTable[] = distinctCourses.map(
+      (courseId: string) => ({
+        courseId,
+        sections: Array.from(
+          new Set(
+            students
+              .filter((s: Student) => s.courseId === courseId)
+              .map((s: Student) => s.sectionName),
+          ),
+        ).map((sectionName: string) => ({
+          sectionName,
+          isAllowedToViewStudentInSection: false,
+          isAllowedToModifyStudent: false,
+          students: students
+            .filter(
+              (s: Student) =>
+                s.courseId === courseId && s.sectionName === sectionName,
+            )
+            .map((s: Student) => ({
+              name: s.name,
+              email: s.email,
+              status: s.joinState,
+              team: s.teamName,
+            })),
+        })),
+      }),
+    );
+
+    return coursesWithSections;
+  }
+
+  getPrivileges(
+    coursesWithSections: SearchStudentsTable[],
+  ): Observable<InstructorPrivilege[]> {
+    return forkJoin(
+      coursesWithSections.map((course: SearchStudentsTable) => {
+        return course.sections.map((section: StudentListSectionData) => {
+          return this.httpRequestService.get(
+            ResourceEndpoints.INSTRUCTOR_PRIVILEGE,
+            {
+              courseid: course.courseId,
+              sectionname: section.sectionName,
+            },
+          );
+        });
+      }).reduce(
+        (acc: Observable<InstructorPrivilege>[], val: Observable<InstructorPrivilege>[]) =>
+        acc.concat(val),
+        [],
```

Comment:
Idk, could this be done so that we can remove the reduce part?
```
  // Gets an array of Observable<InstructorPrivilege>[] via flattening
  const instructorPrivilegesArray = courseWithSections
    .flatMap((course: SearchStudentsTable) => 
      course.sections.map((section: StudentListSectionData)
        => this.http...(req here);
  return forkJoin(instructorPrivilegesArray)
```

Reply:
Also could consider using the `instructorService` service to get the instructor privilege instead of calling `httpRequestService` directly

Reply:
> Idk, could this be done so that we can remove the reduce part?

https://github.com/TEAMMATES/teammates/blob/5f35ee0592e69640fde47c735088f1d6d39fdd6b/tsconfig.json#L27

I would love to use `flatMap` but our project is configured for ES2018 and `flatMap` only ships in [ES2019](https://github.com/tc39/proposals/blob/master/finished-proposals.md). It's easy for us to switch since we're already transpiling, just need to change the tsconfig.

@xpdavid @wkurniawan07 what do you think?

Reply:
Ah right forgot about the ES2019 thing. Sure that works then